docs: Guides for single node benchmarking (#509)
ai-dynamo/dynamo: examples/llm/benchmarking.md

<!--
SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
SPDX-License-Identifier: Apache-2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->

# LLM Deployment Benchmarking Guide

This guide provides detailed steps on benchmarking Large Language Models (LLMs) in single and multi nodes configurations.

> [!NOTE]
> We advice trying out the [LLM Deployment Examples](./README.md) before benchmarking.

## Build the Benchmarking Image

1\. Clone Dynamo repository
```bash
git clone --single-branch --depth=1 -b main https://github.com/ai-dynamo/dynamo.git
```

2\. Build vLLM image
```bash
./container/build.sh
```

> Note: Make sure to run it in the `dynamo` directory cloned.

Author: I think if the user is doing benchmarking, the user should know how Docker works.

  kv-transfer-config: '{"kv_connector":"DynamoNixlConnector"}'
  block-size: 128
  max-model-len: 16384
  remote-prefill: true
  conditional-disagg: true
  max-local-prefill-length: 10
  max-prefill-queue-size: 2
  tensor-parallel-size: 4
  ServiceArgs:
    workers: 1
    resources:
      gpu: 4

PrefillWorker:
  model: neuralmagic/DeepSeek-R1-Distill-Llama-70B-FP8-dynamic
  kv-transfer-config: '{"kv_connector":"DynamoNixlConnector"}'
  block-size: 128
  max-model-len: 4096
  max-num-batched-tokens: 4096
  tensor-parallel-size: 1
  ServiceArgs:
    workers: 4
    resources:
      gpu: 1
```
Key settings:
* **PrefillWorker**: Four separate processes (each with 1 GPU) handle the initial prefill (context embedding) phase.
* **VllmWorker**: One process (using 4 GPUs) generates output tokens (the "decode" phase).
* **kv-transfer-config**: Uses the DynamoNixlConnector for remote KV cache transfer, so that the KV cache can be passed between prefill workers and the generation worker.
* **FP8 Model**: Reduces memory usage for both the model and KV cache, easing remote transfer.
* **Block Size 128**: Batches token processing for more efficient chunk transfers to GPUs.

> 5\. Start disaggregated service

Reviewer: Is it typo?

Author: Do you mean `5\.` vs `5.`? The `\` is intended to escape the `.` so there isn't an added indentation at the numbers, where the lines below do not carry the same indentation. You can see the effect where there is a [typo at `8.`](https://github.com/ai-dynamo/dynamo/blob/6b9713d999c4e3f5a03e9168f900fec2063df7e2/examples/llm/benchmarking.md#disaggregated-multi-node-benchmarking).

    --num-dataset-entries $(($concurrency*12)) \
    --random-seed 100 \
    -- \
    -v \
    --max-threads 256 \
    -H 'Authorization: Bearer NOT USED' \
    -H 'Accept: text/event-stream'

done
```
Key settings:
* **streaming**: Enables token streaming for more realistic performance measurements.
* **concurrency**: Number of simultaneous requests.
* **isl**: Input sequence length.
* **osl**: Output sequence length requested.

## Disaggregated Multi Node Benchmarking

Two H100 80GB x8 node is required for this setup.

With the Dynamo repository and the benchmarking image available, perform the following steps:

1\. Run benchmarking container (node 0 & 1)
```bash
./container/run.sh -it
```
Note: Make sure to run it in the `dynamo` directory cloned.

2\. Download model (node 0 & 1)
```bash
huggingface-cli download neuralmagic/DeepSeek-R1-Distill-Llama-70B-FP8-dynamic
```

3\. Start NATS and ETCD (node 0)
```bash
nats-server -js -p 4222 -m 8222 &
etcd --listen-client-urls http://0.0.0.0:2379 --advertise-client-urls http://0.0.0.0:2379 &
```

4\. Config NATS and ETCD (node 1)
```bash
export NATS_SERVER="nats://<node_0_ip_addr>"
export ETCD_ENDPOINTS="<node_0_ip_addr>:2379"
```

> Note: Node 1 must be able to reach Node 0 over the network for the above services.

Author: I suppose the user knows how compute networking works, i.e. opening ports .... I think this is going to make the guide too lengthy and we should only focus on keys steps needed for benchmarking.

<!--
SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
SPDX-License-Identifier: Apache-2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->

> # LLM Deployment Benchmarking Guide

Author: Will address result interpretation, and enhance comments on settings, on a follow up PR.

This guide provides detailed steps on benchmarking Large Language Models (LLMs) in single and multi nodes configurations.

> [!NOTE]

> > We advice trying out the [LLM Deployment Examples](./README.md) before benchmarking.

Reviewer 2: ```suggestion
> We recommend trying out the [LLM Deployment Examples](./README.md) before benchmarking.
```

Author: Updated

## Build the Benchmarking Image

1\. Clone Dynamo repository
```bash

> git clone --single-branch --depth=1 -b main https://github.com/ai-dynamo/dynamo.git

Reviewer 2: Most other examples assume repo is cloned already - should we continue that and start from build?

Reviewer 3: +1 to this

Author: Sure, updated to guide to start from build.

```

2\. Build vLLM image
```bash
./container/build.sh
```

> Note: Make sure to run it in the `dynamo` directory cloned.

Reviewer 2: ```suggestion
```

Author: Removed

## Disaggregated Single Node Benchmarking

One H100 80GB x8 node is required for this setup.

With the Dynamo repository and the benchmarking image available, perform the following steps:

1\. Run benchmarking container
```bash
./container/run.sh -it
```
Note: Make sure to run it in the `dynamo` directory cloned.

2\. Download model
```bash

> huggingface-cli download neuralmagic/DeepSeek-R1-Distill-Llama-70B-FP8-dynamic

Reviewer 2: do we want to download outside the container for reuse and mount the directory during run? - or can use mount workspace on run in order to get outputs / save model?

Author: Sure, updated guide to advice users to download the model outside the container, and mount the huggingface/hub during run.

```

3\. Start NATS and ETCD
```bash

> nats-server -js -p 4222 -m 8222 &

Reviewer 2: this is starting within the container - most examples start outside the container and use 
```
docker compose -f deploy/docker_compose.yml up -d 
```

Can we normalize that and use it here too?

Reviewer: How can you use docker compose at two nodes solution?

It is slurm or K8S environment and none of these support docker compose, which is strictly single node solution for local deployments. 

You must clearly start depending services explicitly. User must know the ports and network intercase used here to ensure that both nodes can talk with each other.

Author: I normalized this step under "Start NATS and ETCD", and hinted with the docker compose statement.

At the end it is up to the user to start NATS and ETCD under their environment, i.e. maybe they can start multiple containers in a node, or they will have to start it in the head node container.

etcd --listen-client-urls http://0.0.0.0:2379 --advertise-client-urls http://0.0.0.0:2379 &
```

> 4\. Update `/workspace/examples/llm/configs/disagg.yaml` to the following

Reviewer 2: this may be trickier to do within the container - can we have user edit outside the container as first step and then mount workspace?

Reviewer 3: What if we actually save our benchmarking YAML configs to a folder? It would allow users to reproduce quickly. Maybe inside of `examples/` we have a `benchmarks/` dir that also contains a `README`, `graphs/` and `configs`

Author: This is a good idea! The guide is moved under `examples/llm/benchmarks`, and have the graphs and configs for benchmarking contained in the same directory.

Reviewer: It is very good idea to include some configuration examples with short readme about what hardware they are optimal for. The people will ask about new GPUs also so we can later include good configuration example for new GPUs.

Still you need to state assumptions about deployment like we did during answering question about 70B model. The expected average input/output length and the model size. You can reuse answer from the issue: https://github.com/ai-dynamo/dynamo/issues/402#issuecomment-2769181444

Author: The comment on model size is added to `disagg.yaml`, and the input/output token length will be addressed https://github.com/ai-dynamo/dynamo/pull/509#discussion_r2034208231

Processor:
  model: neuralmagic/DeepSeek-R1-Distill-Llama-70B-FP8-dynamic
  router: round-robin

VllmWorker:
  model: neuralmagic/DeepSeek-R1-Distill-Llama-70B-FP8-dynamic
  kv-transfer-config: '{"kv_connector":"DynamoNixlConnector"}'
  block-size: 128
  max-model-len: 3500
  remote-prefill: true
  tensor-parallel-size: 4
  gpu-memory-utilization: 0.95
  disable-log-requests: true
  ServiceArgs:
    workers: 1
    resources:
      gpu: 4

PrefillWorker:
  model: neuralmagic/DeepSeek-R1-Distill-Llama-70B-FP8-dynamic
  kv-transfer-config: '{"kv_connector":"DynamoNixlConnector"}'
  block-size: 128
  max-model-len: 3500
  max-num-batched-tokens: 3500
  tensor-parallel-size: 1
  gpu-memory-utilization: 0.95
  disable-log-requests: true
  ServiceArgs:
    workers: 4
    resources:
      gpu: 1
```
Key settings:
* **PrefillWorker**: Four separate processes (each with 1 GPU) handle the initial prefill (context embedding) phase.
* **VllmWorker**: One process (using 4 GPUs) generates output tokens (the "decode" phase).
* **kv-transfer-config**: Uses the DynamoNixlConnector for remote KV cache transfer, so that the KV cache can be passed between prefill workers and the generation worker.

> * **FP8 Model**: Reduces memory usage for both the model and KV cache, easing remote transfer.

Author: Will address config update on a follow up PR.

* **Block Size 128**: Batches token processing for more efficient chunk transfers to GPUs.

> 

Author: Will address adding explanation on a follow up PR.

> 

Reviewer 2: do we have a scenario that we can add when conditional remote prefill is valuable?

5\. Start disaggregated service
```bash
cd /workspace/examples/llm
dynamo serve graphs.disagg:Frontend -f configs/disagg.yaml 1> disagg.log 2>&1 &
```
Note: Check the `disagg.log` to make sure the service is fully started before collecting performance numbers.

6\. Create and run the benchmarking script
```bash
#/bin/bash

model=neuralmagic/DeepSeek-R1-Distill-Llama-70B-FP8-dynamic

isl=3000
osl=150

for concurrency in 1 2 4 8 16 32 64 128 256; do

  genai-perf profile \
    --model ${model} \
    --tokenizer ${model} \
    --service-kind openai \
    --endpoint-type chat \
    --endpoint /v1/chat/completions \
    --streaming \
    --url http://localhost:8000 \
    --synthetic-input-tokens-mean ${isl} \
    --synthetic-input-tokens-stddev 0 \
    --output-tokens-mean ${osl} \
    --output-tokens-stddev 0 \
    --extra-inputs max_tokens:${osl} \
    --extra-inputs min_tokens:${osl} \
    --extra-inputs ignore_eos:true \
    --concurrency ${concurrency} \
    --request-count $(($concurrency*10)) \
    --warmup-request-count $(($concurrency*2)) \

>     --num-dataset-entries $(($concurrency*12)) \

Author: This thing has a default value of 100
https://github.com/triton-inference-server/perf_analyzer/blob/06df1b7348a17bf87b2f9a3a2b6ae08f9f19ed98/genai-perf/genai_perf/inputs/input_constants.py#L125

We need to take control of it to make sure we are not accidentally sending the "same" prompt over and over again.

>     --num-dataset-entries $(($concurrency*12)) \

Reviewer 4: if prefix-caching is off, does it matter (for non-MoE models) to have a lot of repeated prompts?

>     --num-dataset-entries $(($concurrency*12)) \

Author: I don't think it matters with no caching.

The idea behind is to have a unified script for collecting numbers for this tutorial, we may have to set the prefix-caching flag to on for multi-node setup, which will be drafted soon.

    --random-seed 100 \
    -- \
    -v \
    --max-threads 256 \
    -H 'Authorization: Bearer NOT USED' \
    -H 'Accept: text/event-stream'

done
```
Key settings:
* **streaming**: Enables token streaming for more realistic performance measurements.
* **concurrency**: Number of simultaneous requests.
* **isl**: Input sequence length.
* **osl**: Output sequence length requested.

## Disaggregated Multi Node Benchmarking

> Two H100 80GB x8 node is required for this setup.

Reviewer 2: ```suggestion
Two H100 80GB x8 nodes are required for this setup.
```

With the Dynamo repository and the benchmarking image available, perform the following steps:

1\. Run benchmarking container (node 0 & 1)
```bash
./container/run.sh -it
```
Note: Make sure to run it in the `dynamo` directory cloned.

2\. Download model (node 0 & 1)
```bash
huggingface-cli download neuralmagic/DeepSeek-R1-Distill-Llama-70B-FP8-dynamic
```

3\. Start NATS and ETCD (node 0)
```bash
nats-server -js -p 4222 -m 8222 &
etcd --listen-client-urls http://0.0.0.0:2379 --advertise-client-urls http://0.0.0.0:2379 &
```

4\. Config NATS and ETCD (node 1)
```bash
export NATS_SERVER="nats://<node_0_ip_addr>"
export ETCD_ENDPOINTS="<node_0_ip_addr>:2379"
```
Note: Node 1 must be able to reach Node 0 over the network for the above services.

5\. Update `/workspace/examples/llm/configs/disagg_router.yaml` to the following (node 0 & 1)
```yml
Frontend:
  served_model_name: neuralmagic/DeepSeek-R1-Distill-Llama-70B-FP8-dynamic
  endpoint: dynamo.Processor.chat/completions
  port: 8000

Processor:
  model: neuralmagic/DeepSeek-R1-Distill-Llama-70B-FP8-dynamic
  block-size: 128

>   max-model-len: 16384

Reviewer 2: why dos the processor need a max model len?

Reviewer: The chat processor can enforce length limit before it is sent to workers:
```
  File "/workspace/examples/llm/components/processor.py", line 197, in chat_completions
    async for response in self._generate(raw_request, RequestType.CHAT):
  File "/workspace/examples/llm/components/processor.py", line 109, in _generate
    ) = await self._parse_raw_request(raw_request)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/examples/llm/utils/chat_processor.py", line 74, in _parse_raw_request
    preprocess_result = await processor.preprocess(raw_request)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/examples/llm/utils/chat_processor.py", line 143, in preprocess
    ) = await self.openai_serving._preprocess_chat(
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/dynamo/venv/lib/python3.12/site-packages/vllm/entrypoints/openai/serving_engine.py", line 434, in _preprocess_chat
    prompt_inputs = await self._tokenize_prompt_input_async(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/dynamo/venv/lib/python3.12/site-packages/vllm/entrypoints/openai/serving_engine.py", line 268, in _tokenize_prompt_input
    return next(
           ^^^^^
  File "/opt/dynamo/venv/lib/python3.12/site-packages/vllm/entrypoints/openai/serving_engine.py", line 291, in _tokenize_prompt_inputs
    yield self._normalize_prompt_text_to_input(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/dynamo/venv/lib/python3.12/site-packages/vllm/entrypoints/openai/serving_engine.py", line 183, in _normalize_prompt_text_to_input
    return self._validate_input(request, input_ids, input_text)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/dynamo/venv/lib/python3.12/site-packages/vllm/entrypoints/openai/serving_engine.py", line 246, in _validate_input
    raise ValueError(
ValueError: This model's maximum context length is 1 tokens. However, you requested 13 tokens (8 in the messages, 5 in the completion). Please reduce the length of the messages or completion.
event: error
: a python exception was caught while processing the async generator: ValueError: This model's maximum context length is 1 tokens. However, you requested 13 tokens (8 in the messages, 5 in the completion). Please reduce the length of the messages or completion.
```

  router: kv

Router:
  model-name: neuralmagic/DeepSeek-R1-Distill-Llama-70B-FP8-dynamic
  min-workers: 1

VllmWorker:
  model: neuralmagic/DeepSeek-R1-Distill-Llama-70B-FP8-dynamic
  kv-transfer-config: '{"kv_connector":"DynamoNixlConnector"}'
  block-size: 128
  max-model-len: 3500
  remote-prefill: true
  tensor-parallel-size: 4
  gpu-memory-utilization: 0.95
  disable-log-requests: true
  router: kv

>   enable-prefix-caching: true

Author: I have tried several times, the multi-node setup won't work if this flag is not set to true.

>   enable-prefix-caching: true

Reviewer 4: I think this is because you use kv router. Can you try round-robin or random? KV router needs to emit kv cache events, which is only supported with the "non-naive" (sorry, forget name) allocator.

>   enable-prefix-caching: true

Author: Will try different routing policy on the multi-node guide PR.

  ServiceArgs:

>     workers: 2

Reviewer: Pls use single worker here because for this example it is more optimal. The previous results indicate that it is possible to get 200 tok/sec/gpu for prefill TP1DP12 generate TP4DP1.

Author: Will investigate multi-node config in a follow up PR.

    resources:
      gpu: 4

PrefillWorker:
  model: neuralmagic/DeepSeek-R1-Distill-Llama-70B-FP8-dynamic
  kv-transfer-config: '{"kv_connector":"DynamoNixlConnector"}'
  block-size: 128
  max-model-len: 3500
  max-num-batched-tokens: 3500
  tensor-parallel-size: 1
  gpu-memory-utilization: 0.95
  disable-log-requests: true
  ServiceArgs:

>     workers: 8

Reviewer: Pls use single worker here because for this example it is more optimal. The previous results indicate that it is possible to get 200 tok/sec/gpu for prefill TP1DP12 generate TP4DP1.

The single run of dynamo serve will fail with 12 because it can only split among workers GPUs from single node.

You need two variants of this config:
- the first for head node with 4 workers.
- the second for compute node with 8 PrefillWorkers and none VllmWorkers.

Author: Will address multi-node config in a follow up PR.

    resources:
      gpu: 1
```

> 

Reviewer 2: we should add "key settings" as above and describe the config here - any deltas with single node can be added as key settings to emphasize the delta

Author: Will address multi-node config in a follow up PR.

6\. Start frontend, processor and decode worker (node 0)
```bash
cd /workspace/examples/llm

> sed -i "s/.link(PrefillWorker)//" graphs/disagg_router.py

Reviewer 2: since we are editing configs - why not edit the .py file as well and show the change? 

Also if we use start is this required - start can start individual services as well - @ishandhanani to confirm best practice here.

This should line up with @ishandhanani 's multi-node 405 b example in general

Reviewer 3: I think you should be able to use the instructions here @kthui - https://github.com/ai-dynamo/dynamo/blob/main/examples/llm/multinode-examples.md#single-node-sized-models. It's pretty detailed and does what you're trying to do here I believe

Reviewer 3: Yep - this is a more fine-grained way of doing it. For benchmarking this is a better approach. It also lets you view logs independently on each screen

Author: I think we can update the guide, if we are going to separate the start command of each component, on [the other docs](https://github.com/ai-dynamo/dynamo/tree/main/examples/llm#example-architectures)?

Reviewer: If doc is like from "zero to hero" tutorial, then using ``dynamo serve`` as single command is good approach. This document is for more advanced users, so in my opinion the more details can help them with understanding how each component works.

Author: Will address `dynamo serve` single vs multi commands in a follow up PR.
